helix-term/src/commands.rs

```diff
@@ -1610,13 +1610,47 @@ fn search_impl(
     // it out, we need to add it back to the position of the selection.
     let mut offset = 0;
 
+    // Compare matches without taking text into account
+    // Text is different depending on if its forward or a backward search
+    // So we use start and end to tell them apart
+    let is_match_eq =
+        |m1: &regex::Match, m2: &regex::Match| m1.start() == m2.start() && m1.end() == m2.end();
     // use find_at to find the next match after the cursor, loop around the end
     // Careful, `Regex` uses `bytes` as offsets, not character indices!
     let mut mat = match direction {
         Direction::Forward => regex.find_at(contents, start),
         Direction::Backward => regex.find_iter(&contents[..start]).last(),
     };
 
+    // Find all matches in the document
+    // Then find the current match position inside all matches
+    // We then render that to the status line by setting editor.search_matches
+    let all_matches: Vec<_> = regex.find_iter(contents).collect();
```

Comment:
I'm a little reluctant about this feature in general because it requires collecting all matches in the document. `n` only needs to find the first match after the `start` point and `N` works on a subslice of the document in most cases but with this change, all invocations of `n` and `N` find all matches of the pattern across the doc

Reply:
It can be cached like this https://github.com/helix-editor/helix/commit/07c5e9a9408dd556e9e23864682322a3a882a21c (I can clanup the code) between consecutive n or N usages.

What do you think? I personally find this feature very useful it feels weird when I'm searching a document and I don't know how many elements ahead there is

Reply:
Storing the matches on the doc (preferrably within SearchPosition) looks good. We should be able to avoid running any regex if we already have the set of matches, and we may want to perform the search async so that we don't block the UI when there are many matches https://github.com/helix-editor/helix/issues/2811#issuecomment-1170189784

Reply:
I merged them in one struct. I'll probably need some time to figure out how to make the search async and to use only one regex search, probably we'll need to keep a state in the search like a cursor for where we are at.

Reply:
@the-mikedavis can you review/test https://github.com/helix-editor/helix/commit/efdb69af9e417c7f427ede9c6888864fa1aa885a This commit only makes one regex search, if that works and I can clean it up I'll probably tackle async next

Reply:
I just noticed a bug when the regiester `/` changes the search needs to be also reset, I can fix that later after your next review

Reply:
Bug fixed https://github.com/helix-editor/helix/compare/master...sigmaSd:helix:reg?expand=1 I'll keep pushing updates to that branch